spr/src/commit_message.rs

```diff
@@ -0,0 +1,426 @@
+    }
+
+    #[test]
+    fn test_parse_no_newline_before_body() {
+        assert_eq!(
+            parse_commit_message("No newline before body\nThe body"),
+            CommitMessage {
+                subject: s("No newline before body"),
+                body: s("The body"),
+                trailers: TrailerMap::new(),
+            },
+        );
+    }
+
+    #[test]
+    fn test_parse_subject_and_body() {
+        assert_eq!(
+            parse_commit_message("Subject and body\n\nThe body\nparts"),
+            CommitMessage {
+                subject: s("Subject and body"),
+                body: s("The body\nparts"),
+                trailers: TrailerMap::new(),
+            },
+        );
+    }
+
+    #[test]
+    fn test_parse_subject_and_body_with_paragraphs() {
+        assert_eq!(
+            parse_commit_message(r#"Body with paragraphs
+
+Paragraph1
+ends here.
+
+Paragraph2
+ends here.
+
+Paragraph3
+ends here.
+
+
+"#),
+            CommitMessage {
+                subject: s("Body with paragraphs"),
+                body: s(r#"Paragraph1
+ends here.
+
+Paragraph2
+ends here.
+
+Paragraph3
+ends here."#),
+                trailers: TrailerMap::new(),
+            },
+        );
+    }
+
+    #[test]
+    fn test_parse_single_line_trailers() {
+        assert_eq!(
+            parse_commit_message(r#"Single line trailers
+
+Paragraph1
+ends here.
+
+Foo: FOO1    FOO2    
+Bar:     BAR1 BAR2
+
+"#),
+            CommitMessage {
+                subject: s("Single line trailers"),
+                body: s("Paragraph1\nends here."),
+                trailers: TrailerMap::from( [
+                    ( s("Foo"), vec![ s("FOO1    FOO2") ] ),
+                    ( s("Bar"), vec![ s("BAR1 BAR2") ] ),
+                ] ),
+            },
+        );
+    }
+
+    #[test]
+    fn test_parse_multi_line_trailers() {
+        assert_eq!(
+            parse_commit_message(r#"Multi-line trailers
+
+Body with list:
+
+- foo
+- bar
+- baz
+
+Foo: FOO1
+  FOO2
+Bar:     BAR1
+  BAR2
+
+"#),
+            CommitMessage {
+                subject: s("Multi-line trailers"),
+                body: s("Body with list:\n\n- foo\n- bar\n- baz"),
+                trailers: TrailerMap::from( [
+                    ( s("Foo"), vec![ s("FOO1 FOO2") ] ),
+                    ( s("Bar"), vec![ s("BAR1 BAR2") ] ),
+                ] ),
+            },
+        );
+    }
+
+    #[test]
+    fn test_parse_multiple_trailer_entries() {
+        assert_eq!(
+            parse_commit_message(r#"Multiple trailer entries
+
+The body.
+
+Foo: FOO1
+Bar: BAR1 BAR2
+Foo: FOO2 FOO3
+Bar: BAR3
+
+"#),
+            CommitMessage {
+                subject: s("Multiple trailer entries"),
+                body: s("The body."),
+                trailers: TrailerMap::from( [
+                    ( s("Foo"), vec![ s("FOO1"), s("FOO2 FOO3") ] ),
+                    ( s("Bar"), vec![ s("BAR1 BAR2"), s("BAR3") ] ),
+                ] ),
+            },
+        );
+    }
+
```

Comment:
How about where there are trailers + there's an "errant" trailer as part of the body?

```
subject

body

Incorrectly-Placed-Trailer: Value

Trailer1: v1
Trailer2: v2
```
presumably hte incorrect one is part of body